crates/loader/src/local/mod.rs

```diff
@@ -71,6 +71,17 @@ async fn prepare(
 ) -> Result<Application<CoreComponent>> {
     let info = info(raw.info, &src);
 
+    let id_counts: HashMap<String, u8> =
+        raw.components.iter().fold(HashMap::new(), |mut counts, c| {
+            *counts.entry(c.id.clone()).or_insert(0) += 1;
+            counts
+        });
+    for (id, count) in id_counts {
+        if count > 1 {
```

Comment:
Since this bails at the first duplicate, rather than counting all occurrences, you could instead maintain a set or vector of "seen" IDs and bail as soon as you see the first duplicate, e.g.

```
let mut seen = FavouriteCollection::new();
for c in components {
  let id = c.id;
  if seen.contains(id) {
    bail!
  } else {
    seen.add(id);
  }
}
```

(I know it means using a mutable collection, but if, as suggested, you parcel it out to a separate function then the mutation is safely isolated from the rest of the program.)